book/src/developers/designs/cryptography-key-domains.md

```diff
@@ -0,0 +1,88 @@
+# Cryptography Key Domains
+
+Within Kanidm we have to manage a number of private keys with various cryptograhpic purposes. In the
+current design, we have evolved where for each purposes keys are managed in unique ways. However we
+need to improve this for a number reasons including shared keys for Oauth2 domains and a future
+integration with PKCS11.
+
+## Current System
+
+In the current system we store keys in database records in a variety of bespoke ways for different
+uses. Often this means that keys are associated to the object they are integrated with, such as the
+oauth2 client configuration. However this limits us since we may have users that wish to combine
+multiple oauth2 clients into a single security domain, where access tokens may be exchanged between
+clients for forwarded authentication.
+
+Another example is api-tokens for service accounts. In the past we associated a private key to each
+service-account, however this causes issues with key management. We want to combine all of these
+keys but preserve existing api-token signatures. Currently we have no mechanism to do this.
+
+## Use Cases
+
+- Multiple OAuth2 clients wish to be able to access a shared resource server. `access_tokens` issued
+  to client a or b should be accepted by the resource server.
+- Support key-rotation within a key-domain, where former public keys are retained in a valid state,
```

Comment:
Either way we should keep these for an audit history (or log their signature + public matter in #22) for later forensic purposes.

Reply:
yes, I was planning to. We need to keep them anyway because if we had key A, then roll A -> B, tokens signed by A may still be valid and require a path to validate them.